udunits-1.12.9/src/lib/udunits.3
." $Id: udunits.3,v 1.7 2003/08/29 18:29:55 steve Exp $
." $__Header$
.TH UDUNITS 3 "$Date: 2003/08/29 18:29:55 $" "Printed: \n(yr.\n(mo.\n(dy" "UNIDATA LIBRARY FUNCTIONS"
.SH NAME
udunits, utInit, utScan, utCalendar, utInvCalendar, utConvert, utPrint, utIsTime, utHasOrigin, utClear, utCopy, utScale, utDivide, utInvert, utMultiply, utRaise, utTerm \- Unidata units library
.SH SYNOPSIS
.nf
.ft B
cc -I\fIunidata_inc\fP ...
.sp
#include <udunits.h>
.sp
.ta 4m +\w'const utUnit 'u
int utInit(
	const char	*path);
.sp
int utIsInit();
.sp
int utScan(
	const char	*spec,
	utUnit	*unit);
.sp
int utCalendar(
	double	value;
	utUnit	*unit,
	int	*year,
	int	*month,
	int	*day,
	int	*hour,
	int	*minute,
	float	*second);
.sp
int utInvCalendar(
	int	year,
	int	month,
	int	day,
	int	hour,
	int	minute,
	double	second,
	utUnit	*unit,
	double	*value);
.sp
int utConvert(
	const utUnit	*from,
	const utUnit	*to,
	double	*slope,
	double	*intercept);
.sp
int *utPrint(
	const utUnit	*unit,
	char		**s);
.sp
utUnit *utClear(
	utUnit	*unit);
.sp
int *utIsTime(
	const utUnit	*unit);
.sp
int *utHasOrigin(
	const utUnit	*unit);
.sp
utUnit *utCopy(
	const utUnit	*source,
	utUnit	*dest);
.sp
utUnit *utScale(
	utUnit	*source,
	double	factor,
	utUnit	*result);
.sp
utUnit *utDivide(
	utUnit	*numer,
	utUnit	*denom, 
	utUnit	*result);
.sp
utUnit *utInvert(
	utUnit	*source,
	utUnit	*result);
.sp
utUnit *utMultiply(
	utUnit	*term1,
	utUnit	*term2, 
	utUnit	*result);
.sp
utUnit *utRaise(
	utUnit	*source,
	int	power,
	utUnit	*result);
.sp
void utTerm();
.ft
.fi
.SH DESCRIPTION
.LP
The Unidata units library, \fBudunits\fP, supports conversion of unit
specifications between formatted and binary forms, arithmetic
manipulation of unit specifications, and conversion of values between 
compatible scales of measurement.
.LP
A unit is the amount by which a physical quantity is measured.  For example:
.sp
.TS
center tab(|);
c c
c c.
Physical Quantity|Possible Unit
\_|\_
time|weeks
distance|centimeters
power|watts
.TE
.sp
A unit can have an origin associated with it \-\- in which case
the unit and origin together define a scale.  For example, the
phrase "the temperature is 25 degrees Celsius" specifies a particular point on a
measurement scale; whereas the phrase "the temperature difference is
25 degrees Celsius" specifies a unit with no origin and, hence, no associated
scale.  If not remembered, this subtle distinction can cause problems when
handling units.
.LP
.B utInit()
initializes the units package.  If \fIpath\fP is non-NULL and not empty,
then it specifies a units file containing initializing unit definitions;
otherwise, the environment variable \fBUDUNITS_PATH\fR is checked and,
if it exists and is not empty, then it is assumed to contain the pathname
of the units file;
otherwise, a compile-time default pathname is used.
.sp
The definitions in the units file are read into memory.
This function returns 
0 on success, 
UT_ENOFILE if the units file doesn't exist,
UT_ESYNTAX if the units file contains a syntax error,
UT_EUNKNOWN if the units file contains an unknown specification,
UT_EIO if an I/O error occurred while accessing the units file, and
UT_EALLOC if a memory allocation failure occurred.
.LP
.B utIsInit()
indicates whether or not the units package has been initialized.
.sp
This function returns 
0 if the package hasn't been initialized and !0 if the package has been
initialized.
.LP
.B utScan()
converts the formatted unit specification \fIspec\fP into a binary
unit structure and stores the result in \fIunit\fP.
The binary representation is used for algebraic manipulation.
This function returns 0 on success,
UT_ENOINIT if the package hasn't been initialized,
UT_EINVALID if the unit argument is a null pointer,
UT_EUNKNOWN if the specification contains an unknown unit, and
UT_ESYNTAX if the specification contains a syntax error.
.LP
.B utCalendar()
converts the amount, \fIvalue\fP, of the temporal unit, \fIunit\fP,
into a UTC-referenced date and time 
(see, however, the section on HANDLING TIME).
The reference unit shall be a time unit and have an origin.
This function returns 0 on success,
UT_ENOINIT if the package hasn't been initialized and
UT_EINVALID if the unit structure is not a temporal one.
.LP
.B utInvCalendar()
converts a UTC-referenced date and time
into the amount, \fIvalue\fP, of the temporal unit, \fIunit\fP
(see, however, the section on HANDLING TIME).
The reference unit shall be a time unit and have an origin.
This function returns 0 on success,
UT_ENOINIT if the package hasn't been initialized and
UT_EINVALID if the unit structure is not a temporal one.
.LP
.B utConvert()
returns the coefficients of the Galilean transformation (i.e. y = a*x + b)
necessary to convert the \fIfrom\fP unit into the \fIto\fP unit.
The units must be compatible (i.e., their quotient must be dimensionless).
On successful return, \fIslope\fP and \fIintercept\fP will contain the 
values for the slope and intercept coefficients, respectively.
This function returns
0 on success,
UT_ENOINIT if the package hasn't been initialized,
UT_EINVALID if one of the unit structures is invalid, and
UT_ECONVERT if the units are not convertible.
.LP
.B utPrint()
converts the binary unit structure \fIunit\fP
into a formatted unit specification and stores the string into a static
buffer.  
The argument \fIs\fP is set to point to the static buffer.
The string should not be modified (it may, however, be copied)
and will be overwritten by subsequent calls to this function.
This function returns
0 on success,
UT_ENOINIT if the package hasn't been initialized, and
UT_EINVALID if the unit structure is invalid.
On error, the string argument is set to NULL.
.LP
.B utClear()
clears a unit structure by setting it to the dimensionless value 1.
.LP
.B utIsTime()
returns true if the given unit structure refers to a time unit and false
otherwise.
This function ignores whether or not the unit has an origin.
.LP
.B utHasOrigin()
returns true of the given unit structure has an origin (i.e. defines a
scale) and false otherwise.
.LP
.B utCopy()
copies the unit structure \fIsource\fP to the unit structure
\fIdest\fP.
This function correctly handles the case where the same unit structure
is referenced by the source and destination units.
The address of the destination unit structure is returned.
.LP
.B utScale()
scales the unit structure \fIsource\fP by 
\fIfactor\fP, storing the result in the unit structure \fIresult\fP.
This function correctly handles the case where the same unit structure
is referenced by the source and result units.
The address of the result unit structure is returned.
.LP
.B utInvert()
inverts the unit structure \fIsource\fP, storing the result in
unit structure \fIresult\fP.
Multiplying a unit by its reciprocal yields the
dimensionless value 1.
This function correctly handles the case where the source and result
unit refer to the same structure.
If successful, this function returns the address of the result
unit structure; otherwise, it returns NULL.
.LP
.B utDivide()
divides unit structure \fInumer\fP by unit structure \fIdenom\fP
and stores the result in unit structure \fIresult\fP.
This function correctly handles the case where the same unit structure
is referenced by two or more arguments.
If successful, this function returns the address of the result
unit structure; otherwise, it returns NULL.
.LP
.B utMultiply()
multiplies unit structure \fIterm1\fP by unit structure \fIterm2\fP
and stores the result in unit structure \fIresult\fP.
This function correctly handles the case where the same unit structure
is referenced by two or more arguments.
If successful, this function returns the address of the result
unit structure; otherwise, it returns NULL.
.LP
.B utRaise()
raises the unit structure \fIsource\fP by the power \fPpower\fP,
storing the result in the unit structure \fIresult\fP.
This function correctly handles the case where the same unit structure
is referenced by the source and result units.
If successful, this function returns the address of the result
unit structure; otherwise, it returns NULL.
.LP
.B utTerm()
terminates usage of this package.  In particular, it frees all allocated
memory.  It should be called when the library is no longer needed.
.SH "HANDLING TIME"
.LP
The \fBudunits\fP(3) package uses a mixed Gregorian/Julian calendar
system.
Dates prior to 1582-10-15 are assumed to use the Julian calendar, which
was introduced by Julius Caesar in 46 BCE and is based on a year that 
is exactly 365.25 days long.
Dates on and after 1582-10-15 are assumed to use the Gregorian calendar,
which was introduced on that date and is based on a year that is
exactly 365.2425 days long.
(A year is actually approximately 365.242198781 days long.)
Seemingly strange behavior of the \fBudunits\fP(3) package can
result if a user-given time interval includes the changeover date.
For example, \fButCalendar\fP() and \fButInvCalendar\fP() can be used
to show that 1582-10-15 *preceeded* 1582-10-14 by 9 days.
.SH "EXAMPLES"
.LP
Convert two data sets to
a common unit, subtract one from the other, then
save the result in a (different) output unit:
.sp
.RS +4
.nf
.ta 4m +\w'utUnit 'u +\w'double 'u +4m +4m
if (utInit("") != 0) {
	/* handle initialization error */
} else {
	char	*UnitString1, *UnitString2, *OutputUnitString;
	utUnit	unit1, unit2, OutputUnit;
	...
	if (utScan(UnitString1, &unit1) != 0 || 
		utScan(UnitString2, &unit2) != 0 ||
		utScan(OutputUnitString2, &OutputUnit) != 0) {
		/*
		 * handle decode error
		 */
	} else {
		double	InSlope, InIntercept;
		double	OutSlope, OutIntercept;
		...
		if (utConvert(&unit2, &unit1, &InSlope, 
			&InIntercept) != 0 ||
		    utConvert(&unit1, &OutputUnit, &OutSlope, 
			&OutIntercept) != 0) {
			/*
			 * handle data-incompatibility
			 */
		} else {
			/*
			 * process data using:
			 *	OutputValue = OutSlope*(Data1Value -
			 *		(InSlope*Data2Value + InIntercept)) 
			 *		+ OutIntercept
			 */
		}
	}
	utTerm();
}
.fi
.RE
.LP
the above example could be made more efficient
by testing the returned conversion factors for nearness to 1 and 0 and using
appropriately streamlined processing expressions.
.sp
.LP
Compute a threshold value corresponding to an input data value plus a
user-specified delta (the units of the input data value and delta can
differ):
.sp
.RS +4
.nf
.ta 4m +\w'utUnit 'u +\w'double 'u +4m +4m
char	*input_unit_string, *delta_unit_string;
float	input_value;
utUnit	input_unit, delta_unit;
\&...
if (utScan(input_unit_string, &input_unit) != 0 || 
	utScan(delta_unit_string, &delta_unit) != 0) {
	/*
	 * handle decode error
	 */
} else {
	double	slope, intercept;
	...
	if (utConvert(&delta_unit, &input_unit, &slope, &intercept) != 0) {
		/*
		 * handle units incompatibility
		 */
	} else {
		float	threshold = input_value + slope*delta_value 
				    + intercept;
		...
	}
}
utTerm();
.fi
.RE
.sp
.LP
Compute the number of time intervals from a start time to a 
reference time:
.sp
.RS +4
.nf
#include <stdio.h>
#include <stdlib.h>
#include "udunits.h"

main()
{
    if (utInit("/upc/cur/udunits/etc/udunits.dat") != 0) {
        (void) fputs("utInit() error\n", stderr);
    } else {
        utUnit  timecenters_unit;

        /*
         * NB: The following "time centers" specification gives both
         * the start time (January 1, 1990 at 00:00 UTC) and the
	 * sampling interval (2 minutes).
         */
        if (utScan("2 minutes since 1990-1-1", &timecenters_unit) 
                != 0) {
            (void) fputs("utScan() error\n", stderr);
        } else {
            /*
             * The following reference time is 1 hour later than
             * the above start time.
             */
            int         ref_year        = 1990;
            int         ref_month       = 1;
            int         ref_day         = 1;
            int         ref_hour        = 1;
            int         ref_minute      = 0;
            float       ref_second      = 0;	/* could be `double'
            double      ref_value;

            (void) utInvCalendar(ref_year, ref_month, ref_day,
                                 ref_hour, ref_minute, ref_second,
                                 &timecenters_unit, &ref_value);
            /*
             * Exit successfully if the number of time intervals
             * between the start and reference times is correct.
             */
            if (30 == ref_value)
                exit(0);
        }
    }
    abort();
}
.sp
.fi
.RE
.SH "FORMATTED UNIT SPECIFICATIONS"
.LP
The following are examples of formatted unit specifications that can be
interpreted by the 
.B utScan()
function:
.sp
.RS +4
.nf
10 kilogram.meters/seconds2
10 kg-m/sec2
10 kg\ m/s^2
(PI\ radian)2
degF
100rpm
geopotential meters
33 feet water
.fi
.RE
.LP
A unit is specified as an arbitrary product of constants and unit names
raised to arbitrary integral powers.
Division is indicated by a slash `/'.
Multiplication is indicated by whitespace, a period `.', or a hyphen `-'.
Exponentiation is indicated by an
integer suffix or by the exponentiation operators `^' and `**'.
Parentheses may be used for grouping and disambiguation.
.LP
Arbitrary Galilean transformations (i.e. y = ax + b) are supported.
In particular, temperature and time conversions are correctly handled.
The specification:
.sp
.RS
degF @ 32
.RE
.sp
indicates a Fahrenheit scale with the origin shifted to thirty-two
degrees Fahrenheit (i.e. to zero degrees Celsius).
The Celsius scale is equivalent to the following unit:
.sp
.RS
.nf
1.8 degR @ 273.15
.fi
.RE
.sp
Besides the character `\fB@\fP', the words `\fBafter\fP', `\fBfrom\fP',
`\fBref\fP', and `\fBsince\fP' may also be used.
Note that multiplication takes precedence over origin-shift.
In order of increasing precedence, the operations are origin-shift, division,
multiplication, and exponentiation.
.LP
Units of time are similarly handled.
The specification:
.sp
.RS
.nf
seconds since 1992-10-8 15:15:42.5 -6:00
.fi
.RE
.sp
indicates seconds since October 8th, 1992 at 3 hours, 15 minutes and 42.5
seconds
in the afternoon in the time zone which is six hours to the west of
Coordinated Universal Time (i.e. Mountain Daylight Time).
The time zone specification
can also be written without a colon using one or two-digits (indicating
hours) or three or four digits (indicating hours and minutes).
.LP
.B utScan()
understands most conventional prefixes and abbreviations:
.sp
.TS
center tab(|);
c c c
l l c.
Factor|Prefix|Abbreviation
\_|\_|\_
1e24|yotta|Y
1e21|zetta|Z
1e18|exa|E
1e15|peta|P
1e12|tera|T
1e9|giga|G
1e6|mega|M
1e3|kilo|k
1e2|hecto|h
1e1|deca, deka|da
1e-1|deci|d
1e-2|centi|c
1e-3|milli|m
1e-6|micro|u
1e-9|nano|n
1e-12|pico|p
1e-15|femto|f
1e-18|atto|a
1e-21|zepto|z
1e-24|yocto|y
.TE
.LP
The function
.B utPrint()
always encodes a unit specification one way.  To reduce
misunderstandings, it is recommended that
this encoding style be used as the default.  
In general, a unit is printed in terms of basic units, factors, and
exponents.
Basic units are separated by spaces; and any
exponent directly appends its associated unit.
The above examples would be printed as follows:
.sp
.RS +4
.nf
10 kilogram meter second-2
9.8696044 radian2
0.555556 kelvin @ 255.372
10.471976 radian second-1
9.80665 meter2 second-2
98636.5 kilogram meter-1 second-2
.fi
.RE
.sp
Note that the Fahrenheit unit is encoded as a deviation, in fractional 
kelvins, from an origin at 255.372 kelvin.
.SH "UNITS FILE"
.LP
The units file is a formatted file containing unit definitions and is
used to initialize this package.
It is the first place to look to discover the set of
valid names and symbols (of which there are many \-\-  On October 9, 1992, 
it contained 446 entries).
.LP
The format for the units file is documented internally and the file may be
modified by the user as necessary.  In particular, additional units and
constants may be easily added (including variant spellings of existing
units or constants).
.LP
.SH ENVIRONMENT
.nr in \w'UDUNITS_PATH   '
.IP UDUNITS_PATH \n(inu
If \fButInit()\fP
is called without a pathname argument, and if this environment variable is
non-empty, then its value overrides the default
pathname for the units file.
.SH DIAGNOSTICS
.LP
This package prints
(hopefully) self-explanatory error-messages to standard error.
.SH "SEE ALSO"
.LP
.BR udunits (1).
.SH "BUGS AND RESTRICTIONS"
.LP
.B utScan()
is case-sensitive.  If this causes difficulties, you might
try making appropriate additional entries to the units file.
.LP
Some unit abbreviations in the default units file might seem 
counter-intuitive.
In particular, note the following:
.sp
.RS 4
.nf
.ta \w'Celsius   'u +\w'`newton\' or `N\'   'u +\w'`rad\'   'u
For	Use	Not	Which Instead Means
.sp
Celsius	`Celsius'	`C'	coulomb
gram	`gram'	`g'	<standard free fall>
gallon	`gallon'	`gal'	<acceleration>
radian	`radian'	`rad'	<absorbed dose>
Newton	`newton' or `N'	`nt'	nit (unit of photometry)
.fi
.RE
.SH REFERENCES
.LP
NIST Special Publication 811, 1995 Edition: "Guide for the Use of the
International System of Units (SI)" by Barry N. Taylor.  URL
<http://physics.nist.gov/Divisions/Div840/SI.html>.
.LP
ANSI/IEEE Std 260-1978: "IEEE
Standard Letter Symbols for Units of Measurement".
.LP
ASTM Designation: E 380 \- 85: "Standard for METRIC PRACTICE".
.LP
International Standard (ISO) 2955: "Information processing \-\- 
Representation of SI and other units in systems with limited character sets",
Ref. No. ISO 2955-1983 (E).


udunits-1.12.9/src/udunits/udunits.c
/*
 * $Id: udunits.c,v 1.3 2008/04/04 17:29:08 steve Exp $
 *
 * Test program for the Unidata udunits(3) library.
 */

#ifndef _XOPEN_SOURCE
#  define _XOPEN_SOURCE 500
#endif
#include <limits.h>	/* for _POSIX_MAX_INPUT */
#include <stddef.h>	/* for size_t */
#include <stdio.h>	/* for I/O functions, and NULL */
#include <string.h>	/* for strlen() & memmove() */
#include <errno.h>	/* for errno */
#include <ctype.h>	/* for isspace() */
#include "udunits.h"

#ifdef YYDEBUG
    extern int	utdebug;
#endif

static char	*progname;

#define	ABS(x)		((x) < 0 ? -(x) : (x))


    static void
WriteError(name)
    char	*name;
{
    (void) fprintf(stderr, "%s: Error writing to standard output: ", 
		   name);
    perror("");
}


    static void
ReadError(name)
    char	*name;
{
    (void) fprintf(stderr, "%s: Error reading from standard output: ", 
		   name);
    perror("");
}


/*
 * Get a unit specification.
 */
   static int
GetSpec(prompt, spec, size)
    char	*prompt;
    char	*spec;
    size_t	size;
{
    int		status;
    static char	me[]	= "GetSpec";

    if (fputs(prompt, stdout) == EOF) {
	WriteError(me);
	status	= -1;
    } else if (fgets(spec, size, stdin) == NULL) {
	putchar('\n');
	if (feof(stdin)) {
	    status	=  0;
	} else {
	    ReadError(progname);
	    status	= -1;
	}
    } else {
	/*
	 * Trim whitespace from the specification.
	 */
	char	*start, *stop;

	for (start = spec; *start != 0; ++start)
	    if (!isspace(*start))
		break;
	for (stop = start + strlen(start); stop > start; --stop)
	    if (!isspace(stop[-1]))
		break;
	*stop	= 0;
	(void)memmove((void*)spec, (void*)start, (size_t)(stop-start+1));

	status	= 1;
    }

    return status;
}


main(argc, argv)
    int			argc;
    char		**argv;
{
    int			status;
    int			c;
    char		HaveSpec[_POSIX_MAX_INPUT+1];
    char		WantSpec[_POSIX_MAX_INPUT+1];
    utUnit		HaveUnit, WantUnit;
    static char		*UnitsPath	= NULL;
    extern int		optind;
    extern char		*optarg;

    progname	= argv[0];

    while ((c = getopt(argc, argv, "d")) != -1)
	switch (c) {
	case 'd':
#ifdef YYDEBUG
	    utdebug	= 1;
#endif
	    break;
	default:
	    (void) fprintf(stderr, "%s: Unknown option `%c'\n", progname, c);
	    exit(1);
	}

    if (optind < argc)
	UnitsPath	= argv[optind];

    if (utInit(UnitsPath) != 0) {
	(void) fprintf(stderr, 
		       "%s: Couldn't initialize udunits(3) package\n", 
		       progname);

    } else {
	for (;;) {
	    int		i;
	    double	slope, intercept;

	    i	=  GetSpec("You have: ", HaveSpec, sizeof(HaveSpec));
	    if (i == -1)
		goto failure;
	    if (i == 0)
		goto success;
	    if (HaveSpec[0] == 0)
		continue;
	    if (utScan(HaveSpec, &HaveUnit) != 0) {
		(void) fprintf(stderr, 
			       "%s: Don't recognize `%s'\n", 
			       progname, HaveSpec);
		if (ferror(stderr))
		    goto failure;
	    } else {
		for (;;) {
		    i	= GetSpec("You want: ", WantSpec, 
				      sizeof(WantSpec));
		    if (i == -1)
			goto failure;
		    if (i == 0)
			goto success;
		    if (WantSpec[0] == 0) {
			char	*s;

			(void)utPrint(&HaveUnit, &s);
			if (printf("    Definition: \"%s\"\n", s) < 0) {
			    WriteError(progname);
			    goto failure;
			}
			if (utIsTime(&HaveUnit) && atoi(HaveSpec) != 0) {
			    int	year, month, day, hour, minute;
			    float	second;

			    (void) utCalendar(1.0, &HaveUnit, 
					      &year, &month, &day,
					      &hour, &minute, &second);
			    (void) printf(
			    "    \"%s\" is %d-%02d-%02d %02d:%02d:%g UTC\n",
					  HaveSpec, year, month, day, hour,
					  minute, second);
			}
		    } else if (utScan(WantSpec, &WantUnit) != 0) {
			(void) fprintf(stderr, 
				       "%s: Don't recognize `%s'\n", 
				       progname, WantSpec);
			if (ferror(stderr))
			    goto failure;
			continue;
		    } else {
			i	= utConvert(&HaveUnit, &WantUnit, 
					    &slope, &intercept);
			if (i == UT_ECONVERT) {
			    (void) fprintf(stderr,
					   "%s: Units are incompatible\n", 
					   progname);
			    if (ferror(stderr))
				goto failure;
			} else if (i == UT_EINVALID) {
			    (void) fprintf(stderr, 
					   "%s: A unit is corrupted\n", 
					   progname);
			    if (ferror(stderr))
				goto failure;
			} else {
			    if (intercept == 0) {
				if (printf("    <%s> = <%s>*%g\n", 
					   WantSpec, HaveSpec, slope)
				       < 0)  {
				    WriteError(progname);
				    goto failure;
				}
				if (printf("    <%s> = <%s>/%g\n", 
					   WantSpec, HaveSpec, 1./slope)
				       < 0)  {
				    WriteError(progname);
				    goto failure;
				}
			    } else {
				if (printf("    <%s> = <%s>*%g %s %g\n", 
					   WantSpec, HaveSpec, slope,
					   intercept < 0
						? "-"
						: "+",
					   ABS(intercept)) < 0) {
				    WriteError(progname);
				    goto failure;
				}
				if (printf("    <%s> = <%s>/%g %s %g\n", 
					   WantSpec, HaveSpec, 1./slope,
					   intercept < 0
						? "-"
						: "+",
					   ABS(intercept)) < 0) {
				    WriteError(progname);
				    goto failure;
				}
			    }
			}		/* successful conversion */
		    }			/* "want" decoded */
		    break;
		}			/* "want" loop */
	    }				/* "have" decoded */
	}				/* "have" loop */
    }					/* units package initialized */

    failure:
	status	= 1;
	goto exit;
    success:
	status	= 0;
	goto exit;

    exit:
	utTerm();
	return status;
}
